{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPP1XQeIX7h+rs99A4+HpkC",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/zuegereliane/python-intermediate-inflammation/blob/main/Session3and%204/group1.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Session 3 & 4\n",
        "In the last session, you learned the basic processing steps for RNA sequencing data. Now you will do the next steps in the analysis of transcriptomics data. First, you will figure out which genes are up- or downregulated between two conditions using differential gene expression analysis. Then you will link this to biological pathways with Gene Ontology enrichment.\n",
        "\n",
        "# Background\n",
        "\n",
        "iPSCs are valuable resource to model drug interactions with different tissue and are frequenly used in drug screens to gain insights into drug specificity and toxicity. Tyrosine kinase inhibitors (TKIs) are important drugs in cancer treatment. Unfortunately, certain TKIs may lead to cardiomyopathy as a severe side effect of these drugs.\n",
        "\n",
        "Here, you will compare the transcriptomes of iPSC-derived cardiomyocites treated with two different FDA-approved TKIs to investigate their toxicity and gain insights into the biological pathways potentially leading to cardiotoxicity.\n",
        "- Doxorubicin\n",
        "- Gefinitib\n",
        "\n",
        "# Objectives\n",
        "\n",
        "1. Explore the impact on global gene expression patterns (PCA).\n",
        "2. Identify differentially expressed genes (DEGs) between treatment and control for both drugs.\n",
        "3. Visualized DEGs with volcano plots and heatmaps.\n",
        "4. Perform GO enrichment to determine affected biological pathways.\n",
        "5. Interpret the cardiotoxic potential of these drugs, and what cellular mechansims they might disrupt.\n"
      ],
      "metadata": {
        "id": "zcWYMrTFKrKC"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Setup\n",
        "\n",
        "Run the following cells to set up the necessary packages and download the data. If you wish to use a package which is not in the list below, you will need to install and import it yourself.\n"
      ],
      "metadata": {
        "id": "pF0UXld8OkRi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Install packages which are not in the default environment\n",
        "%pip install scanpy\n",
        "%pip install pydeseq2\n",
        "%pip install gseapy\n"
      ],
      "metadata": {
        "id": "oW5pDPclOoER"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Import packages\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import scanpy as sc\n",
        "import anndata as ad\n",
        "import os\n",
        "import pandas as pd\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "from sklearn.decomposition import PCA\n",
        "from pydeseq2.dds import DeseqDataSet\n",
        "from pydeseq2.ds import DeseqStats\n",
        "import pickle\n",
        "import gseapy\n"
      ],
      "metadata": {
        "id": "BjO3iiXpOrnR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Make data directory if it does not exist\n",
        "os.makedirs(\"data\", exist_ok=True)\n",
        "os.makedirs(\"plots\", exist_ok=True)\n",
        "\n",
        "#Download datasets in the data folder\n",
        "\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session3and4/data/counts_group1.txt -O data/Session3and4/counts_group1.txt\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session3and4/data/metadata_group1.txt -O data/Session3and4/metadata_group1.csv\n"
      ],
      "metadata": {
        "id": "ldz1hfMXSz1J"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Preprocessing RNA seq data\n",
        "\n",
        "Preprocess the data like you learned in the previous session.\n",
        "- How many samples do you have?\n",
        "- How many genes do you have before and after filtering?"
      ],
      "metadata": {
        "id": "R6UReLUTUB-5"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "3Xer_JLpVCKA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Principle component analysis (PCA)\n",
        "\n",
        "Now explore your data with PCA like you did in the previous session.\n",
        "- How do the different drug treatments impact global gene expression patterns?\n",
        "- Which drug do you think has a stronger effect on gene expression?\n"
      ],
      "metadata": {
        "id": "YcIH1LU3VCtH"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "poyMB-5OaRQx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Differential gene expression\n",
        "\n",
        "Now you can perform differential gene expression analysis using the DEseq2 method. You will need to do this twice, comparing each drug with the untreated control condition.\n",
        "- First, you need to filter your count table and metadata to only contain the samples you want to compare. Here's an example of how to do this. Replace the ... with the drug name.\n",
        "\n",
        "```\n",
        "drug = \"...\"  \n",
        "\n",
        "# Filter metadata to contain only drug and control\n",
        "metadata_df_filtered = metadata_df[metadata_df[\"treatment\"].isin([drug, \"Control\"])]\n",
        "# Filter count table by matching index with the filtered metadata\n",
        "counts_filtered = counts[metadata_df_filtered.index]\n",
        "```\n",
        "- Use the filtered data to create a DESeq object, and run the DEseq analysis based on the code cheat sheet.\n",
        "- Look at the results of you analysis. What information do the rows and columns of the DEseq result mean.\n",
        "- Now get the upregulated and downregulated genes from this dataframe. What does it mean for a gene to be up-/downregulated?\n",
        "- Plot the results in a volcano plot. How many genes are differentially expressed?\n",
        "- Repeat the analysis for the second drug. Are there more or less differentially expressed genes? Does this match your expectiations from the PCA analysis?\n"
      ],
      "metadata": {
        "id": "f6F3tDNbaRa4"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "6lv00wlYeScy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Gene Ontology enrichment\n",
        "\n",
        "Now, you can start interpreting the lists of genes you get from the DEseq analysis and learn if there are any biological pathways that are effected by the drug treatment.\n",
        "- Use the cheat sheet to perform a GO enrichment analysis on the upregulated and dowregulated genes of each drug treatment. Which pathways can you see?\n",
        "- Based on this analysis, do you think either of the drugs you tested are cardiotoxic? Which cellular mechanisms do they disrupt?"
      ],
      "metadata": {
        "id": "g-4opEFgeS_a"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "sEuAeer3eZyi"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}